grpc.go

```diff
@@ -56,40 +61,96 @@ type onlineResource struct {
 
 var emptyResp = &gpb.Empty{}
 
-func onlineCPUMem() error {
-	resourceList := []onlineResource{
-		{
-			sysfsOnlinePath: sysfsCPUOnlinePath,
-			regexpPattern:   cpuRegexpPattern,
-		},
-		{
-			sysfsOnlinePath: sysfsMemOnlinePath,
-			regexpPattern:   memRegexpPattern,
-		},
+var onlineCPUMemLock sync.Mutex
+
+const onlineCPUMemWaitTime = 100 * time.Millisecond
+
+const onlineCPUMaxTries = 10
+
+// online resources, nbResources specifies the maximum number of resources to online.
+// If nbResources is -1 then there is no limit, all resources are connected
+// returns the number of resources connected
+func onlineResources(resource onlineResource, nbResources int32) (uint32, error) {
+	files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	if err != nil {
+		return 0, err
 	}
 
-	for _, resource := range resourceList {
-		files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	var count uint32
+	for _, file := range files {
+		matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
 		if err != nil {
-			return err
+			return count, err
 		}
 
-		for _, file := range files {
-			matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
-			if err != nil {
-				return err
-			}
+		if !matched {
+			continue
+		}
 
-			if !matched {
+		onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		status, err := ioutil.ReadFile(onlinePath)
+		if err != nil {
+			// resource cold plugged
+			continue
+		}
+
+		if strings.Trim(string(status), "\n\t ") == "0" {
+			if err := ioutil.WriteFile(onlinePath, []byte("1"), 0600); err != nil {
+				agentLog.WithField("online-path", onlinePath).WithError(err).Errorf("Could not online resource")
 				continue
 			}
+			count++
+		}
 
-			onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
-			ioutil.WriteFile(onlinePath, []byte("1"), 0600)
+		if nbResources > 0 && count == uint32(nbResources) {
```

Comment:
nits: move it to after `count++`. It only makes sense to check when `count` increases.